### PR TITLE
Use GraphListener and FlowExecutionListener to receive events

### PR DESCRIPTION
**Pull request by jglick**
# Description

Uses more recently added listener APIs to avoid tricks with `RunListener`.

No special test coverage; presumably would be covered by existing tests.

@reviewbybees esp. @abayer

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-events/src/main/java/io/jenkins/blueocean/events/PipelineEventListener.java [hunk 1]
```diff
@@ -45,135 +44,146 @@
...
+    private final Map<FlowExecution, String> currentStageId = new WeakHashMap<>();
 
-        private final Run run;
-        private final PubsubBus pubSubBus;
-        private String currentStageName;
-        private String currentStageId;
+    @Override
+    public void onNewHead(FlowNode flowNode) {
+        // test whether we have a stage node
+        if (PipelineNodeUtil.isStage(flowNode)) {
```
Comment (jglick) [inline blueocean-events/src/main/java/io/jenkins/blueocean/events/PipelineEventListener.java]: [for reviewers](https://github.com/jenkinsci/blueocean-plugin/pull/1228/files?w=1)

**Comment by michaelneale** [comment]
cc @vivek for your review

**Comment by michaelneale** [comment]
this is ok with me if we can get one other review. Passes the ATH etc. 

thanks @jglick - yeah tests cover this (should anyway)

**Comment by i386** [comment]
Looks reasonable to me 🐝

**Comment by i386** [comment]
Thanks for this one @jglick 💯

**Pull request merged**